include/albatross/src/models/pic_gp.hpp

```diff
@@ -0,0 +1,845 @@
+
+template <typename GrouperFunction, typename InducingFeatureType,
+          typename FeatureType>
+struct PICGPFit {};
+
+template <typename GrouperFunction, typename FeatureType>
+using PICGroupIndexType =
+    typename std::result_of<GrouperFunction(const FeatureType &)>::type;
+
+template <typename FeatureType> struct PICGroup {
+  // Row indices are into the original set of features into the data
+  // used to form S.
+  Eigen::Index initial_row;
+  Eigen::Index block_size;
+  Eigen::Index block_index;
+  albatross::RegressionDataset<FeatureType> dataset;
+};
+
+template <typename GrouperFunction, typename FeatureType>
+using PICGroupMap = std::map<PICGroupIndexType<GrouperFunction, FeatureType>,
+                             PICGroup<FeatureType>>;
+
+template <typename GrouperFunction, typename InducingFeatureType,
+          typename FeatureType>
+struct Fit<PICGPFit<GrouperFunction, InducingFeatureType, FeatureType>> {
+  using PermutationIndices = Eigen::Matrix<Eigen::Index, Eigen::Dynamic, 1>;
+  using GroupIndexType = PICGroupIndexType<GrouperFunction, FeatureType>;
+  using GroupMap = PICGroupMap<GrouperFunction, FeatureType>;
+
+  std::vector<FeatureType> train_features;
+  std::vector<InducingFeatureType> inducing_features;
+  Eigen::SerializableLDLT train_covariance;
+  Eigen::MatrixXd sigma_R;
+  Eigen::PermutationMatrixX P;
+  std::vector<Eigen::VectorXd> mean_w;
+  Eigen::MatrixXd W;
+  std::vector<Eigen::MatrixXd> covariance_Y;
+  Eigen::MatrixXd Z;
+  BlockDiagonalLDLT A_ldlt;
+  GroupMap measurement_groups;
+  Eigen::VectorXd information;
+  Eigen::Index numerical_rank;
+  std::vector<Eigen::SparseMatrix<double>> cols_Bs;
+  GrouperFunction grouper;
+
+  Fit(){};
+
+  Fit(const std::vector<FeatureType> &features_,
+      const std::vector<InducingFeatureType> &inducing_features_,
+      const Eigen::SerializableLDLT &train_covariance_,
+      const Eigen::MatrixXd &sigma_R_, const Eigen::PermutationMatrixX &P_,
+      const std::vector<Eigen::VectorXd> &mean_w_, const Eigen::MatrixXd &W_,
+      const std::vector<Eigen::MatrixXd> &covariance_Y_,
+      const Eigen::MatrixXd &Z_, const BlockDiagonalLDLT &A_ldlt_,
+      const GroupMap &measurement_groups_, const Eigen::VectorXd &information_,
+      Eigen::Index numerical_rank_,
+      const std::vector<Eigen::SparseMatrix<double>> cols_Bs_,
+      GrouperFunction grouper_)
+      : train_features(features_), inducing_features(inducing_features_),
+        train_covariance(train_covariance_), sigma_R(sigma_R_), P(P_),
+        mean_w(mean_w_), W(W_), covariance_Y(covariance_Y_), Z(Z_),
+        A_ldlt(A_ldlt_), measurement_groups(measurement_groups_),
+        information(information_), numerical_rank(numerical_rank_),
+        cols_Bs(cols_Bs_), grouper(grouper_) {}
+
+  void shift_mean(const Eigen::VectorXd &mean_shift) {
+    ALBATROSS_ASSERT(mean_shift.size() == information.size());
+    information += train_covariance.solve(mean_shift);
```

Comment:
Do we also have to shift the local group values (`mean_w`?). Maybe worth just hard failing here (or removing this?) if it hasn't been tested.